### PR TITLE
SC-680-union-with-lists

### PR DESCRIPTION
**Pull request by potan**

### CODE REVIEW

**Review by xrtm000 — CHANGES_REQUESTED** [review]

File: lang/tests/src/test/scala/com/wavesplatform/lang/compiler/ContractCompilerTest.scala [hunk 1]
```diff
@@ -626,7 +626,14 @@ class ContractCompilerTest extends PropSpec with PropertyChecks with Matchers wi
...
           |    match tx {
           |      case tx:InvokeScriptTransaction =>
-          |        isDefined(tx.payment) && isDefined(tx.feeAssetId)
+          |        isDefined(tx.payment) && isDefined(tx.feeAssetId) &&
+          |        match tx.args[0] {
+          |          case _: Int => false
+          |          case _: String => false
+          |          case _: Boolean => false
+          |          case _: ByteVector => false
+          |          case ll => ll[0] == 1
```
[the remark this answers is not in the document]
Comment (potan) [reply lang/tests/src/test/scala/com/wavesplatform/lang/compiler/ContractCompilerTest.scala:635]: It is compiler patch. Evaluator isn't changed.

[the remark this answers is not in the document]
Comment (xrtm000) [reply lang/tests/src/test/scala/com/wavesplatform/lang/compiler/ContractCompilerTest.scala:635]: Type definition from [Types.scala](https://github.com/wavesplatform/Waves/pull/3293/files/5a873430bed7c7bf74e5f8557f7195c87129774a#diff-c1cb7b911dcd8030428112874a3f129175b0c5e5fc3ea1b36a206192f7c3b17fR266) may be used in runtime (e.g. for match), so it is unsafe. In addition, we specifies fixed context for both runtime and compile time.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Types.scala [hunk 2]
```diff
@@ -263,19 +263,21 @@ object Types {
...
+    val argTypes = UNION(LONG, STRING, BOOLEAN, BYTESTR, LIST(UNION(LONG, STRING, BOOLEAN, BYTESTR)))
     CASETYPEREF(
       "InvokeScriptTransaction",
       addProofsIfNeeded(
         List(
           "dApp"       -> addressOrAliasType,
           "feeAssetId" -> optionByteVector,
           "function"   -> STRING,
-          "args"       -> LIST(UNION(LONG, STRING, BOOLEAN, BYTESTR))
+          "args"       -> LIST(argTypes)
```
[the remark this answers is not in the document]
Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Types.scala:275]: It isn't affect evaluator.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Types.scala [hunk 3]
```diff
@@ -263,19 +263,21 @@ object Types {
     )
   )
 
-  def buildInvokeScriptTransactionType(proofsEnabled: Boolean, version: StdLibVersion) =
+  def buildInvokeScriptTransactionType(proofsEnabled: Boolean, version: StdLibVersion) = {
+    val argTypes = UNION(LONG, STRING, BOOLEAN, BYTESTR, LIST(UNION(LONG, STRING, BOOLEAN, BYTESTR)))
```
[the remark this answers is not in the document]
Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Types.scala:267]: No. Full support requires recursive types.

**Review by xrtm000 — APPROVED** [review]

**Review by mrkraft — APPROVED** [review]

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala [hunk 4]
```diff
@@ -259,10 +260,15 @@ object ExpressionCompiler {
         case _      => None
       }
       matchTypes <- cases.traverse(c => handleCompositeType(p, c.caseType, Some(exprTypes), allowShadowVarName))
+      defaultType = exprTypes match {
+        case ANY => ANY
+        case UNION(tl, _) => UNION(tl.filter(t => !matchTypes.contains(t)), None)
+        case _ => NOTHING
```
Comment (msmolyakov) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:266]: What if union is empty? And what about backward compatibility for v1,2,3?

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Types.scala [hunk 5]
```diff
@@ -135,15 +145,16 @@ object Types {
...
+      case (l1, UNION(l2, _)) => l2.forall(l1 >= _)
+      case (UNION(l1, _), l2) => l1.exists(_ >= l2)
+      case (_, ANY)               => false
       case (_, NOTHING)           => true
       case (NOTHING, _)           => false
       case (LIST(t1), LIST(t2))   => t1 >= t2
       case (TUPLE(types1), TUPLE(types2)) =>
         types1.length == types2.length && (types1 zip types2).forall { case (t1, t2) => t1 >= t2 }
-      case (l1: FINAL, l2: FINAL) => l1.union >= l2.union
+      case (l1: REAL, l2: REAL) => l1 equivalent l2
```
Comment (msmolyakov) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Types.scala:157]: What if extract value from `Any`?
```
func f(a: Any) = value(a)
```

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala [hunk 6]
```diff
@@ -249,17 +249,19 @@ object Parser {
   def matchCaseP(implicit c: fastparse.P[Any]): P[MATCH_CASE] = {
     def checkForGenericAndGetLastPos(t: Type): Either[INVALID, Option[Pos]] =
       t match {
+        case Single(VALID(position, "List"), Some(VALID(_, AnyType(_)))) => Right(Some(position))
         case Single(name, parameter) =>
           parameter
             .toLeft(Some(name.position))
             .leftMap {
-              case VALID(position, v)              => INVALID(position, s"Unexpected generic match type [$v]")
+              case VALID(position, v)              => INVALID(position, s"Unexpected generic match type $t")
```
Comment (msmolyakov) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala:257]: Why not the `[$v]` but more common `$t`?

File: lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala [hunk 7]
```diff
@@ -2056,6 +2056,25 @@ class IntegrationTest extends PropSpec with PropertyChecks with ScriptGen with M
...
 
+  property("Union with single List") {
+    eval("""match (if false then 2 else [3]) { case n: Int => n case a => a[0] }""", version = V4) shouldBe Right(CONST_LONG(3))
+  }
+
+  property("Union with multiple List") {
+    eval("""match (if false then 2 else if true then [3] else ["qqq"]) { case n: Int => n case a => a[0] }""", version = V4) shouldBe Right(CONST_LONG(3))
+  }
+
+  property("Any type") {
```
Comment (msmolyakov) [inline lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala:2067]: Can we try to get item from `Any`?
```
func f(a: Any) = a[0]
```

**Review by msmolyakov — CHANGES_REQUESTED** [review]

File: lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala (hunk 7 above)
Comment (potan) [reply lang/tests/src/test/scala/com/wavesplatform/lang/IntegrationTest.scala:2067]: No? []-operation require List.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Types.scala (hunk 5 above)
Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Types.scala:157]: It works. f(unit) will fail, f(1) == 1.

Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Types.scala:157]: It produce compilation error.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala (hunk 6 above)
Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala:257]: For ```case x:List[Int] => ...``` old code produced error "Unexpected generic match type [Int]", new code produce more informative "Unexpected generic match type List[Int]".

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala (hunk 4 above)
Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:266]: If UNION is empty default case not available.

Comment (potan) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala:266]: I tried, it is compatible.

**Review by msmolyakov — APPROVED** [review]

**Pull request merged**